traits/trait_dict_object.py

```diff
@@ -24,46 +23,461 @@ class TraitDictEvent(object):
+                Will be an empty dict.
+            removed : dict
+                A copy of the dict before it was removed.
+
+        """
+        if self != {}:
+            removed = self.copy()
+            super().clear()
+            self.notifiy(removed=removed)
+
+    def update(self, adict):
+        """ Update the values in the dict by the new dict.
+
+        Parameters
+        ----------
+        adict : dict
+            The dict from which values will be updated into this dict.
+
+        Notes
+        -----
+        Parameters in the notification:
+            added : dict
+                The dict of the item that was added.
+            changed : dict
+                The dict of the item that was changed.
+            removed : dict
+                Will be an empty dict.
+
+        """
+
+        validated_dict, added, changed = self._validate_dict(adict)
+        super().update(validated_dict)
+        self.notifiy(added, changed)
+
+    def setdefault(self, key, value=None):
+        """ Returns the value if key is present in the dict, else creates the
+        key-value pair and returns the value.
+
+        Parameters
+        ----------
+        key : A hashable type.
+            Key to the item.
+
+        Notes
+        -----
+        Parameters in the notification:
+            added : dict
+                The dict of the item that was added, notification is fired
+                only if the key was absent.
+            changed : dict
+                The dict of the item that was changed.
+            removed : dict
+                Will be an empty dict.
+
+        """
+
+        added = None
+        if key not in self:
+            key = self.validate_key(key)
+            value = self.validate_value(value)
+            added = {key: value}
+
+        result = super().setdefault(key, value)
+
+        if added is not None:
+            self.notifiy(added)
+
+        return result
+
+    def pop(self, key, default_value=None):
+        """ Remove the key from the dict if present and return the key-value
+        pair. If key is absent, the default value is returned and the dict
+        is left unmodified.
+
+        Parameters
+        ----------
+        key : A hashable type.
+            Key to the dict item.
+
+        default_value : any
+            Value to return if key is absent.
+
+        Notes
+        -----
+        Parameters in the notification:
+            added : dict
+                Will be an empty dict
+            changed : dict
+                The dict of the item that was changed
+            removed : dict
+                The dict of the item that was removed, notification fired only
+                if the key was present.
+
+        """
+        if key in self:
+            removed = {key: self[key]}
+            result = super().pop(key)
+            self.notifiy(removed=removed)
+            return result
+        else:
+            return default_value
+
```

Comment:
The original default value in `TraitDictObject.pop` is `Undefined`. `Undefined` is used as a way to identify when the default value is not provided. If the default value is not provided and the key is not found, we should still get a KeyError:
```
>>> x = {}
>>> x.pop("c")  # fail!
...
KeyError: "c"
```

See #987